prov/efa/src/efa_user_info.c

```diff
@@ -354,6 +364,36 @@ int efa_user_info_alter_rxr(struct fi_info *info, const struct fi_info *hints)
 
 			info->domain_attr->mr_mode |= FI_MR_HMEM;
 
+			support_atomic = true;
+			if (hmem_ops[FI_HMEM_CUDA].initialized == true) {
+				/*
+				  We are making an assumption here that the remote we are performing atomic
+				  operations on is the same as us. This might not be true for atomics over
+				  client/server model using EFA.  We do not currently have any users trying
+				  to do this, so we are not worried about this assumption yet.
+				*/
+				fi_param_get_bool(NULL, "hmem_cuda_use_gdrcopy", &hmem_cuda_use_gdrcopy);
+				fi_param_get_bool(NULL, "hmem_cuda_enable_xfer", &hmem_cuda_enable_xfer);
+				support_atomic = hmem_cuda_enable_xfer && !hmem_cuda_use_gdrcopy;
+			}
```

Comment:
I think a boolean `atomics_supported` is a good candidate for a field in our `efa_hmem_info` struct, having it initialized in `efa_hmem_init_cuda()` and queried here instead. Unless it's too cumbersome to grab `efa_domain->hmem_info[FI_HMEM_CUDA]` from whatever's in scope here.

Reply:
My initial implementation put `atomics_supported` in in `efa_hmem_info`.  I later removed it b/c the check it would take wasn't worth it IMO.  Each IOV can have a different HMEM IFACE.  That means we would need a loop on the send path for atomics like this:

```
for (iov in msg) {
   get hmem iface
   check if hmem iface supports atomics
   if it doesn't fail the entire send.
}
```

Which I think is more than what we want.  In the worst case, a sender can include hundreds of IOV's which would give us hundreds of if checks.  Those IOV's can all be pointed at the same memory, or different memory.  I figured we would trust the sender to use our provider as the info structure specifies.

Reply:
The second issue with this approach is `efa_user_info_alter_rxr()` gets set in `fi_getinfo()` and `efa_hmem_info_init_all() -> efa_hmem_info_init_cuda()` gets called in `fi_domain()` which happens after.